server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,465 @@
+     * or the last known status if not {@link DecommissionStatus#FAILED}
+     * Once the status is updated, it tries to exclude to-be-decommissioned cluster manager nodes from Voting Configuration
+     *
+     * @param decommissionAttribute register decommission attribute in the metadata request
+     * @param listener register decommission listener
+     */
+    public void startDecommissionAction(
+        final DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        // register the metadata with status as DECOMMISSION_INIT as first step
+        clusterService.submitStateUpdateTask("decommission [" + decommissionAttribute + "]", new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                // validates if correct awareness attributes and forced awareness attribute set to the cluster before starting action
+                validateAwarenessAttribute(decommissionAttribute, awarenessAttributes, forcedAwarenessAttributes);
+                DecommissionAttributeMetadata decommissionAttributeMetadata = currentState.metadata().decommissionAttributeMetadata();
+                // check that request is eligible to proceed
+                ensureEligibleRequest(decommissionAttributeMetadata, decommissionAttribute);
+                decommissionAttributeMetadata = new DecommissionAttributeMetadata(decommissionAttribute);
+                logger.info("registering decommission metadata [{}] to execute action", decommissionAttributeMetadata.toString());
+                return ClusterState.builder(currentState)
+                    .metadata(Metadata.builder(currentState.metadata()).decommissionAttributeMetadata(decommissionAttributeMetadata))
+                    .build();
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                logger.error(
+                    () -> new ParameterizedMessage(
+                        "failed to start decommission action for attribute [{}]",
+                        decommissionAttribute.toString()
+                    ),
+                    e
+                );
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                DecommissionAttributeMetadata decommissionAttributeMetadata = newState.metadata().decommissionAttributeMetadata();
+                assert decommissionAttribute.equals(decommissionAttributeMetadata.decommissionAttribute());
+                decommissionClusterManagerNodes(decommissionAttributeMetadata.decommissionAttribute(), listener);
+            }
+        });
+    }
+
+    private synchronized void decommissionClusterManagerNodes(
```

Comment:
The `synchronized` block here and below doesn't guarantee complete thread safety for the full execution since the executing thread completes execution shortly after making the async call and doesn't block on the callback

Reply:
But keeping it `synchronized ` might help a little here as the first job of this method is to match predicate before sending transport call for exclusion. It would be good iff only one thread does the predicate matching at a time

Reply:
I would request we do a follow up to get rid of synchronising concurrent request handling by handling the state transitions

Reply:
Yes, upcoming PR would attempt to deal with concurrency more gracefully